Sources/protoc-gen-swift/MessageStorageDecision.swift

```diff
@@ -0,0 +1,190 @@
+// Sources/protoc-gen-swift/MessageGenerator.swift - Per-message logic
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/master/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+
+import SwiftProtobufPluginLibrary
+
+// The file attempts to isolate the decisions around when to use heap based
+// storage vs. inlining it into the value times. At the moment the decicions
+// are entirely based on field counts and language requires, but this could be
+// revised in the future to better account for the actually *real* memory
+// impacts.
+
+/// Wraps the calcuation of the "cost" of fields.
+///
+/// As mentioned in the file comment, these numbers can be revised in the futute
+/// to compute a real stack/heap cost if desired.
+fileprivate enum FieldCost {
+  /// Of a repeated field.
+  static let repeated = 1
+  /// Of a map field.
+  static let map = 1
+
+  /// Of "Plan Old Data" (ints, floating point) field that is 64 bits
+  static let singlePOD32 = 1
+  /// Of "Plan Old Data" (ints, floating point) field that is 64 bits
+  static let singlePOD64 = 1
+  /// Of bool field.
+  static let singleBool = 1
+  /// Of a `string` field.
+  static let singleString = 1
+  /// Of a `bytes` field.
+  static let singleBytes = 1
+
+  /// A single Message field where the message in question uses storgage.
+  static let singleMessageFieldUsingStorage = 1
+
+  static func estimate(_ field: FieldDescriptor) -> Int {
+    guard field.label != .repeated else {
+      // Repeated fields don't count the exact types, just fixed costs.
+      return field.isMap ? FieldCost.map : FieldCost.repeated
+    }
+
+    switch field.type {
+    case .bool:
+      return 1
+    case .int32, .sint32, .uint32, .fixed32, .sfixed32, .float, .enum:
+      return FieldCost.singlePOD32
+    case .int64, .sint64, .uint64, .fixed64, .sfixed64, .double:
+      return FieldCost.singlePOD64
+    case .string:
+      return FieldCost.singleString
+    case .bytes:
+      return FieldCost.singleBytes
+    case .group, .message:
+      return analyse(descriptor: field.messageType!).costAsField
+    }
+  }
+}
+
+/// Maxium computed cost of a Message's fields allow before it uses Storage.
+fileprivate let totalFieldCostRequiringStorage = 17
+
+/// The result of analysis, if the message should use heap storage and the
+/// cost of the message when used as a field in other messages.
+fileprivate struct AnalyseResult {
```

Comment:
Not-strongly-held opinion: `AnalyzeResult` and `analyze(descriptor:)` for consistency, since we probably use American English spelling throughout other APIs 🤷

Reply:
Done